File: complete_code/cpp/ball_tracker.cpp
#include "ball_tracker.h"

void BallTracker::startProxies() {
  motion_proxy_ = new AL::ALMotionProxy(IP, PORT);
  posture_proxy_ = new AL::ALRobotPostureProxy(IP, PORT);
  memory_proxy_ = new AL::ALMemoryProxy(IP, PORT);

  motion_proxy_->wakeUp();
  posture_proxy_->goToPosture("StandInit", 0.5);
  motion_proxy_->setAngles("HeadPitch", 0.3, 0.1);
}

void BallTracker::trackBallWithHead(int x, int y, int width, int height) {
  float hor_dist_to_centre = x - 0.5 * width;
  float ver_dist_to_centre = y - 0.5 * height;
  if (std::abs(hor_dist_to_centre) > 15) {
    int direction = std::signbit(hor_dist_to_centre) ? 1 : -1;
    motion_proxy_->changeAngles("HeadYaw", 0.03*direction, 0.1);
  }
  if (std::abs(ver_dist_to_centre) > 15) {
    int direction = std::signbit(ver_dist_to_centre) ? 1 : -1;
    motion_proxy_->changeAngles("HeadPitch", -0.03*direction, 0.1);
  }
}

void BallTracker::moveToBall(int x, int y, int width, int height) {
  float robot_x = 1.f;
  float robot_y = 0.f;
  float theta = 0.f;

  float hor_dist_to_centre = x - 0.5 * width;
  if (std::abs(hor_dist_to_centre) > 10) {
    // If ball not in centre, turn towards the ball
    int dir = std::signbit(x - 0.5 * width) ? 1 : -1;
    theta = 0.3 * dir;
  }
  if (y > (height / 2 + 15)) {
    // If close to ball, walk slower
    robot_x = 0.5;
  }
  motion_proxy_->moveToward(robot_x, robot_y, theta);
}

void BallTracker::considerKicking(int x, int y, int width, int height) {
  if ((y > height - 50) && (std::abs(x - 0.5 * width) < 25)) {
    // If ball is close enough and in front of the robot
    // Check if its head is correct to kick the ball
    float pitch_value = memory_proxy_->getData("Device/SubDeviceList/HeadPitch/Position/Sensor/Value");
    float yaw_value = memory_proxy_->getData("Device/SubDeviceList/HeadYaw/Position/Sensor/Value");
    if (pitch_value < 0.27 || pitch_value > 0.33) {
      motion_proxy_->setAngles("HeadPitch", 0.3, 0.1);
    }
    else if (yaw_value < -0.1 or yaw_value > 0.1) {
      motion_proxy_->setAngles("HeadYaw", 0.0, 0.1);
    }
    else {
      bool side = std::signbit(x - 0.5 * width) ? false : true;
      motion_proxy_->stopMove();
      kick(side);
    }
  } else {
    // Ball is not close or in right position
    moveToBall(x, y, width, height);
  }
}

void BallTracker::searchForBall() {
  motion_proxy_->stopMove();
  motion_proxy_->setAngles("HeadYaw", head_search_yaw_[head_search_index_], 0.1);
  motion_proxy_->setAngles("HeadPitch", head_search_pitch_[head_search_index_], 0.1);
  ++head_search_index_;
  // TODO: not hardcode the 8
  if (head_search_index_ > 8) {
    head_search_index_ = 0;
  }
}

void BallTracker::track() {
  OrangeBall ball_detector;
  ball_detector.setRobotIpAndPort(IP, PORT);
  ball_detector.startRobotCamera(1);
  startProxies();
  cv::Mat img;
  while (true) {
    ball_detector.getImgFromRobot(img);
    Ball ball = ball_detector.detectOrangeBall(img);

    // If ball found
    if (!ball.r == std::numeric_limits<int>::infinity()) {
      // Resert search for ball variables
      times_no_ball_ = 0;
      head_search_index_ = 0;
      if (tracking_method_ == TrackingMethod::head) {
        trackBallWithHead(ball.x, ball.y, img.cols, img.rows);
      } else if (tracking_method_ == TrackingMethod::walk) {
        moveToBall(ball.x, ball.y, img.cols, img.rows);
      } else if (tracking_method_ == TrackingMethod::kick) {
        considerKicking(ball.x, ball.y, img.cols, img.rows);
      }
      // Draw circles on the ball
      cv::circle(img, cv::Point(ball.x, ball.y), ball.r, cv::Scalar(0, 255, 0), 2);
      cv::circle(img, cv::Point(ball.x, ball.y), 2, cv::Scalar(0, 0, 255), 3);
    } else {
      // No ball found
      if (times_no_ball_ > 5) {
        searchForBall();
      }
      ++times_no_ball_;
    }
    // Show image
    cv::imshow("Orange ball detector", img);
    // Quit with esc, Q or q
    int key = cv::waitKey(1);
    if (key == 27 || key == 'Q' || key == 'q') break;
  }

  cv::destroyAllWindows();
  ball_detector.stopRobotCamera();

  // Let the robot sit when done
  motion_proxy_->stopMove();
  motion_proxy_->rest();
}

void BallTracker::kick(bool side=false) {
    posture_proxy_->goToPosture("StandInit", 0.5);
    bool isAbsolute = true;

    if (side) {
      // Kick right
      std::vector<std::string> names;
      names.reserve(14);
      names.push_back("RShoulderRoll");
      names.push_back("RShoulderPitch");
      names.push_back("LShoulderRoll");
      names.push_back("LShoulderPitch");
      names.push_back("RHipRoll");
      names.push_back("RHipPitch");
      names.push_back("RKneePitch");
      names.push_back("RAnklePitch");
      names.push_back("RAnkleRoll");
      names.push_back("LHipRoll");
      names.push_back("LHipPitch");
      names.push_back("LKneePitch");
      names.push_back("LAnklePitch");
      names.push_back("LAnkleRoll");

      AL::ALValue angleLists;
      angleLists.arraySetSize(14);
      angleLists[0]  = AL::ALValue::array(-0.3f);
      angleLists[1]  = AL::ALValue::array(0.4f);
      angleLists[2]  = AL::ALValue::array(1.0f);
      angleLists[3]  = AL::ALValue::array(1.0f);
      angleLists[4]  = AL::ALValue::array(0.0f);
      angleLists[5]  = AL::ALValue::array(-0.4f, -0.2f);
      angleLists[6]  = AL::ALValue::array(0.95f, 1.5f);
      angleLists[7]  = AL::ALValue::array(-0.55f, -1.0f);
      angleLists[8]  = AL::ALValue::array(0.2f);
      angleLists[9]  = AL::ALValue::array(0.0f);
      angleLists[10] = AL::ALValue::array(-0.4f);
      angleLists[11] = AL::ALValue::array(0.95f);
      angleLists[12] = AL::ALValue::array(-0.55f);
      angleLists[13] = AL::ALValue::array(0.2f);

      AL::ALValue timeLists;
      timeLists.arraySetSize(14);
      timeLists[0]  = AL::ALValue::array(0.5f);
      timeLists[1]  = AL::ALValue::array(0.5f);
      timeLists[2]  = AL::ALValue::array(0.5f);
      timeLists[3]  = AL::ALValue::array(0.5f);
      timeLists[4]  = AL::ALValue::array(0.5f);
      timeLists[5]  = AL::ALValue::array(0.4f, 0.8f);
      timeLists[6]  = AL::ALValue::array(0.4f, 0.8f);
      timeLists[7]  = AL::ALValue::array(0.4f, 0.8f);
      timeLists[8]  = AL::ALValue::array(0.4f);
      timeLists[9]  = AL::ALValue::array(0.5f);
      timeLists[10] = AL::ALValue::array(0.4f);
      timeLists[11] = AL::ALValue::array(0.4f);
      timeLists[12] = AL::ALValue::array(0.4f);
      timeLists[13] = AL::ALValue::array(0.4f);

      motion_proxy_->angleInterpolation(names, angleLists, timeLists, isAbsolute);

      AL::ALValue names_alvalue = AL::ALValue::array("RShoulderRoll", "RHipPitch", "RKneePitch", "RAnklePitch", "LShoulderPitch");
      angleLists.clear();
      angleLists.arraySetSize(5);
      angleLists[0]  = AL::ALValue::array(1.0f);
      angleLists[1]  = AL::ALValue::array(-0.7f);
      angleLists[2]  = AL::ALValue::array(1.05f);
      angleLists[3]  = AL::ALValue::array(-0.5f);
      angleLists[4]  = AL::ALValue::array(0.2f);

      timeLists.clear();
      timeLists.arraySetSize(5);
      timeLists[0]  = AL::ALValue::array(0.1f);
      timeLists[1]  = AL::ALValue::array(0.1f);
      timeLists[2]  = AL::ALValue::array(0.1f);
      timeLists[3]  = AL::ALValue::array(0.1f);
      timeLists[4]  = AL::ALValue::array(0.1f);
      motion_proxy_->angleInterpolation(names_alvalue, angleLists, timeLists, isAbsolute);

      names_alvalue = AL::ALValue::array("RHipPitch","RKneePitch", "RAnklePitch");
      angleLists.clear();
      angleLists.arraySetSize(3);
      angleLists[0]  = AL::ALValue::array(-0.5f);
      angleLists[1]  = AL::ALValue::array(1.1f);
      angleLists[2]  = AL::ALValue::array(-0.65f);

      timeLists.clear();
      timeLists.arraySetSize(3);
      timeLists[0]  = AL::ALValue::array(0.25f);
      timeLists[1]  = AL::ALValue::array(0.25f);
      timeLists[2]  = AL::ALValue::array(0.25f);
      motion_proxy_->angleInterpolation(names_alvalue, angleLists, timeLists, isAbsolute);
    } else {
      // Kick right
      std::vector<std::string> names;
      names.reserve(14);
      names.push_back("LShoulderRoll");
      names.push_back("LShoulderPitch");
      names.push_back("RShoulderRoll");
      names.push_back("RShoulderPitch");
      names.push_back("LHipRoll");
      names.push_back("LHipPitch");
      names.push_back("LKneePitch");
      names.push_back("LAnklePitch");
      names.push_back("LAnkleRoll");
      names.push_back("RHipRoll");
      names.push_back("RHipPitch");
      names.push_back("RKneePitch");
      names.push_back("RAnklePitch");
      names.push_back("RAnkleRoll");

      AL::ALValue angleLists;
      angleLists.arraySetSize(14);
      angleLists[0]  = AL::ALValue::array(0.3f);
      angleLists[1]  = AL::ALValue::array(0.4f);
      angleLists[2]  = AL::ALValue::array(-0.5f);
      angleLists[3]  = AL::ALValue::array(1.0f);
      angleLists[4]  = AL::ALValue::array(0.0f);
      angleLists[5]  = AL::ALValue::array(-0.4f, -0.2f);
      angleLists[6]  = AL::ALValue::array(0.95f, 1.5f);
      angleLists[7]  = AL::ALValue::array(-0.55f, -1.0f);
      angleLists[8]  = AL::ALValue::array(-0.2f);
      angleLists[9]  = AL::ALValue::array(0.0f);
      angleLists[10] = AL::ALValue::array(-0.4f);
      angleLists[11] = AL::ALValue::array(0.95f);
      angleLists[12] = AL::ALValue::array(-0.55f);
      angleLists[13] = AL::ALValue::array(-0.2f);

      AL::ALValue timeLists;
      timeLists.arraySetSize(14);
      timeLists[0]  = AL::ALValue::array(0.5f);
      timeLists[1]  = AL::ALValue::array(0.5f);
      timeLists[2]  = AL::ALValue::array(0.5f);
      timeLists[3]  = AL::ALValue::array(0.5f);
      timeLists[4]  = AL::ALValue::array(0.5f);
      timeLists[5]  = AL::ALValue::array(0.4f, 0.8f);
      timeLists[6]  = AL::ALValue::array(0.4f, 0.8f);
      timeLists[7]  = AL::ALValue::array(0.4f, 0.8f);
      timeLists[8]  = AL::ALValue::array(0.4f);
      timeLists[9]  = AL::ALValue::array(0.5f);
      timeLists[10] = AL::ALValue::array(0.4f);
      timeLists[11] = AL::ALValue::array(0.4f);
      timeLists[12] = AL::ALValue::array(0.4f);
      timeLists[13] = AL::ALValue::array(0.4f);

      motion_proxy_->angleInterpolation(names, angleLists, timeLists, isAbsolute);

      AL::ALValue names_alvalue = AL::ALValue::array("LShoulderRoll", "LHipPitch", "LKneePitch",
                                                     "LAnklePitch");
      angleLists.clear();
      angleLists.arraySetSize(4);
      angleLists[0]  = AL::ALValue::array(1.0f);
      angleLists[1]  = AL::ALValue::array(-0.7f);
      angleLists[2]  = AL::ALValue::array(1.05f);
      angleLists[3]  = AL::ALValue::array(-0.5f);

      timeLists.clear();
      timeLists.arraySetSize(4);
      timeLists[0]  = AL::ALValue::array(0.1f);
      timeLists[1]  = AL::ALValue::array(0.1f);
      timeLists[2]  = AL::ALValue::array(0.1f);
      timeLists[3]  = AL::ALValue::array(0.1f);
      motion_proxy_->angleInterpolation(names_alvalue, angleLists, timeLists, isAbsolute);

      names_alvalue = AL::ALValue::array("LHipPitch","LKneePitch", "LAnklePitch");
      angleLists.clear();
      angleLists.arraySetSize(3);
      angleLists[0]  = AL::ALValue::array(-0.5f);
      angleLists[1]  = AL::ALValue::array(1.1f);
      angleLists[2]  = AL::ALValue::array(-0.65f);

      timeLists.clear();
      timeLists.arraySetSize(3);
      timeLists[0]  = AL::ALValue::array(0.25f);
      timeLists[1]  = AL::ALValue::array(0.25f);
      timeLists[2]  = AL::ALValue::array(0.25f);
      motion_proxy_->angleInterpolation(names_alvalue, angleLists, timeLists, isAbsolute);
    }


    posture_proxy_->goToPosture("StandInit", 0.5);
    motion_proxy_->setAngles("HeadPitch", 0.3, 0.1);
}

File: complete_code/python/ball_tracker.py
#!/usr/bin/env python2

import cv2
import itertools
import numpy as np

from naoqi import ALProxy

from orange_ball import create_video_connection, get_img_from_robot, detect_orange_ball

from kick import kick

# Robot name as shown on its head followed by .local
IP = "jerry.local"
PORT = 9559

# TODO: add separate kick
# Tracking method: 0: head, 1: walk to ball
TRACKING_METHOD = 1

HEAD_SEARCH_YAW =   [0, -1,   -2,   -1,   0, 1,   2,   1,   0]
HEAD_SEARCH_PITCH = [0, -0.3, -0.5, -0.3, 0, 0.3, 0.5, 0.3, 0]

HEAD_SEARCH = list(itertools.product(HEAD_SEARCH_YAW, HEAD_SEARCH_PITCH))
NUMBER_HEAD_SEARCH = len(HEAD_SEARCH)

head_search_index = 0

def start_proxies():
    '''
    Start the motion and posture proxy and let the robot stand up.
    '''
    motion_proxy  = ALProxy("ALMotion", IP, PORT)
    posture_proxy = ALProxy("ALRobotPosture", IP, PORT)
    memory_proxy  = ALProxy("ALMemory", IP, PORT)

    motion_proxy.wakeUp()
    posture_proxy.goToPosture("StandInit", 0.5)
    motion_proxy.setAngles("HeadPitch", 0.3, 0.1)

    return motion_proxy, posture_proxy, memory_proxy


def track_ball_with_head(motion_proxy, x, y, width, height):
    '''
    Track the ball with the robots head by changing the head yaw and pitch.
    http://doc.aldebaran.com/2-1/naoqi/motion/control-joint-api.html#ALMotionProxy::changeAngles__AL::ALValueCR.AL::ALValueCR.floatCR
    '''
    hor_dist_to_centre = x - 0.5 * width
    ver_dist_to_centre = y - 0.5 * height
    if np.abs(hor_dist_to_centre) > 15:
        motion_proxy.changeAngles("HeadYaw", -0.03*np.sign(hor_dist_to_centre), 0.1)
    if np.abs(ver_dist_to_centre) > 15:
        motion_proxy.changeAngles("HeadPitch", 0.03*np.sign(ver_dist_to_centre), 0.1)


def reset_head_to_centre(motion_proxy):
    '''
    Reset the head angles to zero.
    http://doc.aldebaran.com/2-1/family/robots/joints_robot.html
    '''
    motion_proxy.setAngles("HeadYaw", 0, 0.1)
    motion_proxy.setAngles("HeadPitch", 0, 0.1)


def move_to_ball(motion_proxy, x, y, width, height):
    '''
    Move towards the ball.
    http://doc.aldebaran.com/2-1/naoqi/motion/control-walk-api.html#ALMotionProxy::moveToward__floatCR.floatCR.floatCR
    '''
    robot_x     = 1.0
    robot_y     = 0.0
    theta       = 0.0
    hor_dist_to_centre = x - 0.5 * width
    if np.abs(hor_dist_to_centre) > 10:
        # If ball not in centre, turn towards the ball
        theta = -0.3 * np.sign(x - 0.5 * width)
    # If close to ball, slowly walk forward
    if y > height / 2 + 15: robot_x = 0.5
    motion_proxy.moveToward(robot_x, robot_y, theta)


def kick_ball(motion_proxy, posture_proxy, x, width):
    '''
    Kick the ball.
    '''
    motion_proxy.stopMove()
    posture_proxy.goToPosture("StandInit", 0.5)
    kick(motion_proxy, 1.0 * np.sign(x - 0.5 * width))
    posture_proxy.goToPosture("StandInit", 0.5)
    motion_proxy.setAngles("HeadPitch", 0.3, 0.1)


def search_for_ball(motion_proxy):
    '''
    Search for the ball by moving the robots head.
    '''
    global head_search_index
    motion_proxy.stopMove()
    # Set head to next command
    head_search_commands = HEAD_SEARCH[head_search_index]
    motion_proxy.setAngles("HeadYaw", head_search_commands[0], 0.1)
    motion_proxy.setAngles("HeadPitch", head_search_commands[1], 0.1)
    head_search_index += 1
    if head_search_index > NUMBER_HEAD_SEARCH - 1:
        head_search_index = 0


def main():
    # Start connection with the robot
    video_device, capture_device = create_video_connection(IP, PORT, 1)
    motion_proxy, posture_proxy, memory_proxy = start_proxies()
    times_no_ball = 0
    while (True):
        img = get_img_from_robot(video_device, capture_device)

        x, y, r = detect_orange_ball(img)

        if x is not None:
            if TRACKING_METHOD == 0:
                track_ball_with_head(motion_proxy, x, y, img.shape[1], img.shape[0])
            elif TRACKING_METHOD == 1:
                move_to_ball(motion_proxy, x, y, img.shape[1], img.shape[0])
                # If ball close enough and in front of the robot, check if its head is correct an kick the ball
                if y > img.shape[0] - 50 and np.abs(x - 0.5 * img.shape[1]) < 25:
                    # Check if head is correct
                    pitch_value = memory_proxy.getData("Device/SubDeviceList/HeadPitch/Position/Sensor/Value")
                    yaw_value = memory_proxy.getData("Device/SubDeviceList/HeadYaw/Position/Sensor/Value")
                    if pitch_value < 0.27 or pitch_value > 0.33:
                        motion_proxy.setAngles("HeadPitch", 0.3, 0.1)
                    elif yaw_value < -0.1 or yaw_value > 0.1:
                        motion_proxy.setAngles("HeadYaw", 0.0, 0.1)
                    else:
                        kick_ball(motion_proxy, posture_proxy, x, img.shape[1])
            # Draw the outer circle
            cv2.circle(img, (x, y), r, (0, 255, 0), 2)
            # Draw the center of the circle
            cv2.circle(img, (x, y), 2, (0, 0, 255), 3)

            # Reset head search index
            head_search_index = 0
            # Reset times that ball was not seen
            times_no_ball = 0
        else:
            if TRACKING_METHOD == 0:
                # Reset head to centre
                reset_head_to_centre(motion_proxy)
            elif TRACKING_METHOD == 1:
                if times_no_ball > 5:
                    search_for_ball(motion_proxy)
                times_no_ball += 1
        # Show image
        cv2.imshow("Orange ball detector", img)

        # Quit with q
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

    cv2.destroyAllWindows()

    video_device.unsubscribe(capture_device)

    motion_proxy.stopMove()
    motion_proxy.rest()


if __name__ == "__main__":
    main()
